ee/katc/config.go

```diff
@@ -0,0 +1,123 @@
+package katc
+
+import (
+	"context"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"log/slog"
+	"runtime"
+
+	"github.com/osquery/osquery-go"
+	"github.com/osquery/osquery-go/plugin/table"
+)
+
+/*
+Open qs:
+- Should we go with the EAV approach rather than with columns? Look at how dataflatten does it
+
+TODOs:
+- Need to do queryContext filtering
+*/
+
+type katcTableType struct {
+	name     string
+	dataFunc func(ctx context.Context, path string, query string, columns []string, slogger *slog.Logger) ([]map[string][]byte, error)
+}
+
+const (
+	sqliteTableType    = "sqlite"
+	indexedDBTableType = "indexeddb"
+)
+
+func (ktt *katcTableType) UnmarshalJSON(data []byte) error {
+	var s string
+	err := json.Unmarshal(data, &s)
+	if err != nil {
+		return fmt.Errorf("unmarshalling string: %w", err)
+	}
+
+	switch s {
+	case sqliteTableType:
+		ktt.name = sqliteTableType
+		ktt.dataFunc = sqliteData
+		return nil
+	case indexedDBTableType:
+		ktt.name = indexedDBTableType
+		return errors.New("indexeddb is not yet implemented")
+	default:
+		return fmt.Errorf("unknown table type %s", s)
+	}
+}
+
+type dataProcessingStep struct {
+	name           string
+	processingFunc func(ctx context.Context, data []byte, slogger *slog.Logger) ([]byte, error)
+}
+
+const (
+	snappyDecodeProcessingStep               = "snappy"
+	structuredCloneDeserializeProcessingStep = "structured_clone"
+)
+
+func (d *dataProcessingStep) UnmarshalJSON(data []byte) error {
+	var s string
+	err := json.Unmarshal(data, &s)
+	if err != nil {
+		return fmt.Errorf("unmarshalling string: %w", err)
+	}
+
+	switch s {
+	case snappyDecodeProcessingStep:
+		d.name = snappyDecodeProcessingStep
+		d.processingFunc = snappyDecode
+		return nil
+	case structuredCloneDeserializeProcessingStep:
+		d.name = structuredCloneDeserializeProcessingStep
+		d.processingFunc = structuredCloneDeserialize
+		return nil
+	default:
+		return fmt.Errorf("unknown data processing step %s", s)
+	}
+}
+
+type katcTableConfig struct {
+	Type                katcTableType        `json:"type"`
```

Comment:
`type` implies something about the type of table, but I think this refers to the source? I'm also not sold on `Path` or `Columns`, they feel specific to `sqlite`.

I wonder if we can take a page from databases and treat this as a connect url string. `sqlite:path` or similar. Each source type would need to define it's own schema there.

Reply:
Yeah, I think `source` is a better name than `type`! I'll rename.

I think `path` should always apply -- e.g. path to JSON file, or indexeddb file, or sqlite file. I'm still waffling a bunch on columns.

Reply:
Maybe `filepath` is better/clearer? I am also not opposed to doing a connect url string approach, that feels like it could be tidy 🙂

Reply:
What if it was a network call?

Reply:
I guess we could overload `path`. It's an arbitrary string parsed by source. :shrug: mostly it comes down to whether you'd like `source` and `path` or `source://path`

Reply:
More thoughts --

I like the idea of replacing `path` with a `connection_string` or `dsn`. That does get a little weird with supporting wildcards -- we'd have to 1) parse the connection string to extract the filepath portion, then 2) glob based on that part, then 3) re-construct the connection string to use it. But we're doing the second two steps anyway, so might as well?

I _think_ I like keeping source in a separate field regardless of what we decide for ^, since it makes config unmarshalling simpler and more self-contained -- but I will think on it more and see how it feels after updating `path`!

Reply:
> re-construct the connection string to use it.

Do we really need that? I think it'll depend a lot on the `type` in question. Like, imagine:
```
type: sqlite
dsn: file:///Users/%/...?options
```

If the dsn starts with file, URL parse it and hand it to glob. But I don't think we need to reconstruct, because I think the options are parsed by our katc type, not the underlying implementation. (practically speaking, maybe they're passed along, but I think that's going to be type specific)

Reply:
I don't think we'd do this, but I can imagine a curl style thing. `https://host/path?validate=false` or something, that validate might signal to the type thing to disable tls validation, but it's not going to be part of the remote URL.

Reply:
Renamed `type` to `source` -- `path` to `dsn` to come later once I sort through some of the other parts!

Reply:
Okay, I've ended up with `source_type` (for `sqlite`, `leveldb`, etc) and `source` (filepath with wildcard to db, network call, etc). I think this is a lot clearer in the code too, e.g. with the `sourceData` struct.